### ISSUE TITLE
Integrating X-LoRA

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/peft#1472]
Hello all, and thank you for your great work!

Earlier this week, we announced [X-LoRA](https://arxiv.org/abs/2402.07148), a flexible MoE approach for LoRA adapters. We implement deep layer- and token-wise scalings for multiple LoRA adapters and provide an implementation (https://github.com/EricLBuehler/xlora) that enables straightforward application to any model to which `peft` LoRA adapters may be applied to. This offers the possibility to orchestrate at a much finer level, that is, to achieve new combinations of adapter layers. This results in never-before-seen per-token deep layer-wise combinations of parameters to solve specific tasks. Sample weights are provided at: https://huggingface.co/lamm-mit/x-lora with examples in protein science in the paper. 

Would you be interested to perhaps integrate X-LoRA into `peft`? I would be happy to work on this if there is interest from you and the community.

**Comment by BenjaminBossan** [comment]
Thanks for bringing this to our attention, this could be an interesting addition. I just had a quick glance at the repo but it looks quite solid to me.

I think something we should pay attention to if we want to integrate it to PEFT is the API. Ideally, we would like to be able to initiate an X-LoRA model (or load a trained one) just based on a config using the usual PEFT/transformers methods. That way, we could ensure that users can make use of X-LoRA without running custom code. I'm not sure how feasible that is, but it would be great to have.

**Comment by EricLBuehler** [comment]
I think this would be feasible. The primary entry point for converting to X-LoRA is a function `add_xlora_to_model`. It already takes a `PreTrainedModel` and we have a config class that could easily be ported. Specifically, would adding a new model type class where I would port the mechanics from `add_xlora_to_model` to the mapping make sense?

**Comment by BenjaminBossan** [comment]
> would adding a new model type class where I would port the mechanics from `add_xlora_to_model` to the mapping make sense?

Do you mean a new model class on the level of `PeftModel`, or something on the level of `LoraModel`? If the latter, yes, I think that would be a good idea here. Maybe you could just start with a quick draft PR and we see from there? Often it's easier to discuss once we can see the code.

**Comment by EricLBuehler** [comment]
I have an implementation that is nearly ready for a draft PR on my fork. However, after looking at how `PeftModel` initializes its base model, I think a modification to the instantiation call for the `PeftModel` base model may be necessary for `xLoRAModel` to be on the same level as `LoraModel`.

This is because in the `xlora` library we create a `PeftModel` which needs to be modified based on the `xLoRAConfig` (loading adapters, injecting methods and hooks, etc.). For example, we override the `save_pretrained` method to save adapters and the classifier in specific locations.

Perhaps one solution is to pass a reference to the `PeftModel` when creating the `xLoRAModel`? This way, the necessary methods can be overridden, and the `xLoRAModel` would be on the same level as `LoraModel`. Do you think that this method of passing a reference would be a good idea, or is there another way?

**Comment by BenjaminBossan** [comment]
It would be much easier to discuss possible solutions once I see how the code looks like :)

**Comment by EricLBuehler** [comment]
Ok, sounds good! I will submit a draft PR after I complete some small testing.

**Comment by A11en0** [comment]
It appears that the inference process is very slow when using four LoRA experts. In fact, it is significantly slower than four times the single LoRA inference time. Could you please explain why for me, thanks very much!

**Comment by BenjaminBossan** [comment]
@A11en0 Thanks for trying this out so quickly after it was merged :)

Do you have some code to replicate the issue? What times did you measure?

**Comment by A11en0** [comment]
When utilizing Xlora with four experts, processing 2000 samples—each averaging 500 in length—requires approximately 10 hours.

**Comment by A11en0** [comment]
```yaml
### model
model_name_or_path: Mistral-7B-Instruct-v0.2

### method
stage: sft
do_train: true
finetuning_type: xlora
lora_rank: 32
lora_alpha: 32
lora_target: q_proj,k_proj,v_proj,o_proj
saved_lora_adapters: project/outputs/lora_DI,project/outputs/lora_ED,project/outputs/lora_EM,project/outputs/lora_SM
use_trainable_adapters: false

### dataset
dataset: jellyfish
template: jellyfish
cutoff_len: 1024
overwrite_cache: true
preprocessing_num_workers: 16

### output
output_dir: outputs/xlora
logging_steps: 10
save_steps: 500
plot_loss: true
overwrite_output_dir: true

### train
per_device_train_batch_size: 4
gradient_accumulation_steps: 4
learning_rate: 6.0e-5
num_train_epochs: 3.0
lr_scheduler_type: cosine
warmup_ratio: 0.1
bf16: true
ddp_timeout: 180000000
save_total_limit: 10
load_best_model_at_end: true

### eval
val_size: 0.1
per_device_eval_batch_size: 4
eval_strategy: steps
eval_steps: 500

## wandb
report_to: wandb
```

above is my yaml configure file.

**Comment by A11en0** [comment]
```python
        ...
        model = AutoModelForCausalLM.from_pretrained(model_weights).to("cuda")

        model.config.use_cache = False   # for xlora
        if lora_weights:
            model = PeftModel.from_pretrained(model, lora_weights)
        ....
```

**Comment by A11en0** [comment]
When I use a single lora, it needs about only 5 hours. Here is my Xlora config:

``yaml
use_trainable_adapters: true
softmax_temperature: 0.1
```

and others are default.

**Comment by BenjaminBossan** [comment]
So in your case, running X-LoRA with 4 adapters is roughly twice as slow as running LoRA with a single adapter, right? That looks indeed a bit excessive, as most of the time should be spent on the base model, which is the same in each case.

To test this further, I ran a small benchmark to check the overhead of X-LoRA. This benchmark uses a rather small model, `bloomz-1b1`, as the X-LoRA overhead should be expected to be larger the smaller the base model is. Cache was deactivated. For inference, I found this:

- base model: 0.14 sec
- LoRA model: 0.17 sec
- X-LoRA model: 1.10 sec

So there is indeed a quite significant overhead, at least for this "small" model. Not sure if @EricLBuehler has some insights into what the reasons could be.

I attached the notebook with the measurements. Download and change "txt" to "ipynb" to run it.

[xlora-timing.txt](https://github.com/user-attachments/files/16267568/xlora-timing.txt)

**Comment by A11en0** [comment]
Thanks for your quick reply! I have executed the Jupyter notebook you provided and reviewed the source codes of xlora. I noticed that xlora performs two forward passes in the `pre_forward()` function, which should increase inference time by at least 2 times compared to a normal lora model. However, this does not clarify why your results show that the time taken by xlora is only slight more than lora. Could you assist me with this matter once more?

**Comment by YangS03** [comment]
Hi, I also observe **a severe progressive slowdown** when using XLora with PEFT.

Setup:

```python

# Init
from peft import XLoraConfig, get_peft_model

lora_config = XLoraConfig(
    task_type="CAUSAL_LM",
    hidden_size=token_size,
    xlora_depth=4,
    xlora_size=128,
    adapters={
        "0": "...",  # language adapter 1
        "1": "...",  # language adapter 2
    },
)
self.vlm.language_model = get_peft_model(self.vlm.language_model, lora_config)

# Generation
output: CausalLMOutputWithPast = self.vlm.generate(
    input_ids=input_ids,
    attention_mask=attention_mask,
    pixel_values=pixel_values,
    use_cache=False,
    output_hidden_states=False,
    return_dict_in_generate=False,
)
```

Observation during generation:

At the beginning GPU utilization is high(~25%), but as decoding continues, GPU utilization steadily decreases(~1%).

VRAM usage stays stable (no cache growth), but tokens/sec keeps dropping.

**Iteration time per batch grows dramatically (from ~15s to 40–45s per step after only a few dozen generations).**

Examples from logs:

```
<|im_start|>system
You are a helpful assistant.<|im_end|>
<|im_start|>user
<image>
Please describe this image in general. Directly provide the description, do not include prefix like "This image depicts". <|im_end|>
<|im_end|>assistant
 The image captures a snowy landscape, likely a ski resort or mountain area, with a snowboarder in the foreground. The snowboarder is dressed in a grey jacket and blue jeans, and is riding a white snowboard. The background features a snow-covered slope with a ski lift visible in the distance, and a line of tall, dark evergreen trees on the left side. The sky is overcast, and the overall atmosphere is serene and cold.
  0%|▎         | 1/585 [00:15<2:30:16, 15.44s/it]Index(['index', 'image', 'answer', 'question'], 
....

<|im_start|>system
You are a helpful assistant.<|im_end|>
<|im_start|>user
<image>
Please describe this image in general. Directly provide the description, do not include prefix like "This image depicts". <|im_end|>
<|im_end|>assistant
 The image shows a Siamese cat with a blue collar standing on the rim of a toilet bowl. The cat has a light brown and white coat with darker stripes on its face and ears. The toilet is white, and the cat's paws are black. In the background, there is a bathroom setting with a towel hanging on the wall, a plant, and a cabinet. The lighting in the room is soft, and the overall atmosphere is calm and domestic.
  1%|██▎      | 8/585 [05:32<7:02:15, 43.91s/it]Index(['index', 'image', 'answer', 'question'], dtype='object')

```

Has anyone else reproduced this?

**Comment by YangS03** [comment]
After debugging I confirmed that forward hooks are leaking across generations. I added a counter to inspect hooks before/after each call to .generate:

```python
def count_hooks(mod):
    pre, post = 0, 0
    for _, m in mod.named_modules():
        pre  += len(getattr(m, "_forward_pre_hooks", {}) or {})
        post += len(getattr(m, "_forward_hooks", {}) or {})
    return pre, post

print("[DBG] HOOKS before:", count_hooks(self.language_model))

outputs = self.language_model.generate(
    inputs_embeds=input_embeds,
    attention_mask=attention_mask,
    generation_config=generation_config,
    output_hidden_states=output_hidden_states,
    **generate_kwargs,
)

cleanup_xlora_pre_hooks(self.language_model)
```

Observed output:

```
dynamic ViT batch size: 3
[DBG] HOOKS before: (141120, 0)
Setting `pad_token_id` to `eos_token_id`:151645 for open-end generation.
[DBG] HOOKS after : (168336, 0)

dynamic ViT batch size: 7
[DBG] HOOKS before: (168336, 0)
Setting `pad_token_id` to `eos_token_id`:151645 for open-end generation.
[DBG] HOOKS after : (187656, 0)

dynamic ViT batch size: 7
[DBG] HOOKS before: (187656, 0)
Setting `pad_token_id` to `eos_token_id`:151645 for open-end generation.
[DBG] HOOKS after : (211008, 0)
```

So each generation call adds tens of thousands of pre-hooks that never get removed, which explains the steadily worsening performance (high CPU usage, low GPU utilization).

## Temporary Cleanup Function

As a workaround I wrote a cleanup utility that scans all modules for stale X-LoRA scalings_injection_hook entries and removes them:

```python
import functools

def cleanup_xlora_pre_hooks(model, verbose=True):
    cleaned = 0
    for _, m in model.named_modules():
        d = getattr(m, "_forward_pre_hooks", None)
        if not isinstance(d, dict):
            continue
        for k, cb in list(d.items()):
            is_xlora = (
                isinstance(cb, functools.partial)
                and getattr(cb.func, "__name__", "") == "scalings_injection_hook"
            )
            if is_xlora:
                try:
                    d.pop(k, None)
                    cleaned += 1
                except Exception:
                    pass
    if verbose and cleaned:
        print(f"[XLORA] cleaned {cleaned} stale pre_hooks")
```

This stops the growth, but the underlying issue is that `_enable_peft_forward_hooks` registers new pre-hooks on every generation without guaranteeing they are removed.

**Comment by BenjaminBossan** [comment]
@YangS03 Thanks for reporting this error and for investigating the cause. I think it should be fixed via #2761. If you have the opportunity, please check if that branch solves your original issue. In my quick tests, I no longer observed any slowdowns when calling `generate` multiple times.

**Comment by YangS03** [comment]
Thanks, the issue is resolved!
